src/marks/density.js

```diff
@@ -11,44 +12,113 @@ const defaults = {
     return create("svg:g")
         .call(applyIndirectStyles, this, scales, dimensions)
         .call(applyTransform, this, scales)
-        .call(g => g.selectAll("path")
-          .data(contourDensity()
-              .x(X ? i => X[i] : constant(cx))
-              .y(Y ? i => Y[i] : constant(cy))
-              .size([width, height])
-              .bandwidth(bandwidth)
-              .thresholds(thresholds)
-            (index))
+        .call(g => g.selectAll()
+          .data(Array.from(index, i => [i]))
           .enter()
           .append("path")
-            .attr("d", geoPath()))
-      .node();
+            .call(applyDirectStyles, this)
+            .call(applyGroupedChannelStyles, this, channels)
+            .attr("d", ([i]) => path(contours[i])))
+        .node();
   }
 }
 
 export function density(data, {x, y, ...options} = {}) {
   ([x, y] = maybeTuple(x, y));
   return new Density(data, {...options, x, y});
 }
+
+function densityInitializer(options, bandwidth, thresholds, f, s) {
+  return initializer(options, function(data, facets, channels, scales, dimensions) {
+    const X = channels.x?.scale ? valueof(channels.x.value, scales[channels.x.scale]) : channels.x?.value;
+    const Y = channels.y?.scale ? valueof(channels.y.value, scales[channels.y.scale]) : channels.y?.value;
+    const W = channels.weight?.value;
+    const Z = channels.z?.value;
+    const {z} = this;
+    const [cx, cy] = applyFrameAnchor(options, dimensions);
+    const {width, height} = dimensions;
+    const newFacets = [];
+    const contours = [];
+    const newChannels = Object.entries(channels).filter(([key]) => key !== "x" && key !== "y" && key !== "weight").map(([key, d]) => [key, {...d, value: []}]);
+    if (f) newChannels.push(["fill", {value: [], scale: "color"}]);
+    if (s) newChannels.push(["stroke", {value: [], scale: "color"}]);
+    let max = 0, maxn = 0;
+    const density = contourDensity()
+      .x(X ? i => X[i] : constant(cx))
+      .y(Y ? i => Y[i] : constant(cy))
+      .weight(W ? i => W[i] : 1)
+      .size([width, height])
+      .bandwidth(bandwidth)
+      .thresholds(thresholds);
+
+    // First pass: seek the maximum density across all facets and series; memoize for performance.
+    const memo = [];
+    thresholds = [];
+    for (const [facetIndex, facet] of facets.entries()) {
+      newFacets.push([]);
+      for (const index of Z ? groupZ(facet, Z, z) : [facet]) {
+        const c = density(index);
+        const d = c[c.length - 1];
+        if (d) {
+          if (d.value > max) {
+            max = d.value;
+            maxn = c.length;
+            thresholds = c.map(d => d.value);
+          }
+          memo.push({facetIndex, index, c, top: d.value});
+        }
+      }
+    }
+
+    // Second pass: generate contours with the thresholds derived above
+    density.thresholds(thresholds);
+    for (const {facetIndex, index, c: memoc, top} of memo) {
+      const c = top < max ? density(index) : memoc;
```

Comment:
I’m not sure how much this caching is helping; I think most commonly, only one series (or facet-series) will be cacheable as having the max threshold. All the others will need to be recomputed. I guess that’s still helpful in the common case of only a single series and single facet, but I suspect we could make this significantly faster if d3.contourDensity had a way to compute and store the blurred grid, and then compute individual contours for a particular threshold afterward. That way, we wouldn’t have to recompute the underlying grid when computing new contours, and we could cache individual contours rather than sets.